# EBay – Exploring Used Car Sales Data

## 1. Introduction

In this project, we’ll explore and clean a dataset of used car listings from eBay Kleinanzeigen, the German branch of eBay's classified section. The dataset contains information about car listings including price, brand, model, mileage, and other attributes. It has been sampled and intentionally "dirtied" to simulate real-world data cleaning challenges, making it ideal for practicing data wrangling with pandas in a Jupyter environment.

## Data Dictionary

* `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.
* `name` - Name of the car.
* `seller` - Whether the seller is private or a dealer.
* `offerType` - The type of listing
* `price` - The price on the ad to sell the car.
* `abtest` - Whether the listing is included in an A/B test.
* `vehicleType` - The vehicle Type.
* `yearOfRegistration` - The year in which the car was first registered.
* `gearbox` - The transmission type.
* `powerPS` - The power of the car in PS.
* `model` - The car model name.
* `kilometer` - How many kilometers the car has driven.
* `monthOfRegistration` - The month in which the car was first registered.
* `fuelType` - What type of fuel the car uses.
* `brand` - The brand of the car.
* `notRepairedDamage` - If the car has a damage which is not yet repaired.
* `dateCreated` - The date on which the eBay listing was created.
* `nrOfPictures` - The number of pictures in the ad.
* `postalCode` - The postal code for the location of the vehicle.
* `lastSeenOnline` - When the crawler saw this ad last online.


## 2. Exploring the Data

The aim of this project is to clean the data and analyze the included used car listings. So let's start by importing the libraries we need, and reading the dataset into pandas.

In [1]:
# importing libraries
import numpy as np
import pandas as pd

In [2]:
autos = pd.read_csv("autos.csv",encoding="Latin-1")
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [3]:
# Display dataframe summary information
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

From the output above, we observe that:

* The dataset contains `20` columns, most of which are strings.
* There are a few columns with null values, but no columns have more than ~20% null values.
* The column names use **camelcase** instead of Python's preferred **snakecase**, which means we can't just replace spaces with underscores.


## 3. Cleaning Column Names

Let's convert the column names from camelcase to snakecase, and reword some of the column names to be more descriptive.

In [4]:
# Display the column names before renaming
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [5]:
# Rename columns to 'snake_case' for easier access and readability
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
                 'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
                 'odometer', 'registration_month', 'fuel_type', 'brand',
                 'unrepaired_damage', 'date_created', 'pictures_number', 'postal_code',
                 'last_seen']

autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage         date_created  pictures_number  \
0     peugeot              nein  2016-03-26 00:00:00                0   
1         bmw              nein  2016-04-04 00:00:00                0   
2  volkswagen              nein  2016-03-26 00:00:00                0   
3       smart              nein  2016-03-12 00:00:00                0   
4        ford              nein  2016-04-01 00:00:00                0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50




## 4. Initial Exploration and Cleaning

Let's perform some basic data exploration before determining what other cleaning tasks we need to do.



In [6]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-29 23:42:13  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

               date_created  pictures_number   postal_code  \
count                 50000          50000.0  50000.000000   
unique                   76              NaN           NaN   
top     2016-04-03 00:00:00              NaN           NaN   
freq                   1946              NaN           NaN   
mean                    NaN              0.0  50813.627300   
std                     NaN              0.0  25779.747957   
min                     NaN              0.0   1067.000000   
25%                     NaN              0.0  30451.000000   
50%                     NaN              0.0  49577.000000   
75%                     NaN              0.0  71540.000000   
max                     NaN              0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN


From the descriptive statistics table, there are a number of things worth noting:

* The `seller` and `offer_type` columns mostly contain the same values, thus we can safely remove them from the DataFrame.
* The `price` and `odometer` columns contain numeric values stored as text, which can be cleaned and converted.
* `registration_year` minimum value is 1000 and maximum value is 9999 which is not possible.
* `power_ps` minimum value is 0 maximun value is ~17K horsepower which again is not possible till date. ref
* `registration_month` minimum value is 0 and maximum value is 12, which is again not possible as there are only 12 months in a year
* There seems to be something wrong with `pictures_number` column as all of its statistics(except count) is zero.
  

In [7]:
autos['price'] = autos['price'].astype(str).str.replace('$', '').str.replace(',', '').astype(int)

autos['odometer'] = autos['odometer'].astype(str).str.replace('km', '').str.replace(',', '').astype(int)

autos.rename({"odometer": "odometer_km"}, axis=1, inplace=True)

* Converted `price` and `odometer` from text to numeric values.
* Renamed the `odometer` column to `odometer_km` to make it look more descriptive.

In [8]:
autos = autos.drop(["seller", "offer_type", "pictures_number"], axis='columns')
autos.columns

Index(['date_crawled', 'name', 'price', 'ab_test', 'vehicle_type',
       'registration_year', 'gearbox', 'power_ps', 'model', 'odometer_km',
       'registration_month', 'fuel_type', 'brand', 'unrepaired_damage',
       'date_created', 'postal_code', 'last_seen'],
      dtype='object')

* Removed `seller, offer_type, pictures_number` columns before we go any further.

## 5.Exploring the Odometer and Price Columns

Let's continue exploring the data, specifically looking for suspicious data. We'll start by analyzing the odometer_km and price columns using their minimum and maximum values, then remove any outliers that look unrealistically high or low.

We'll use:
* `Series.unique().shape` to see how many unique values
* `Series.describe()` to view min/max/median/mean etc
* `Series.value_counts()`, with some variations:
* chained to `.head()` if there are lots of values.
* Because `Series.value_counts()` returns a series, we can use `Series.sort_index()` with `ascending= True or False` to view the highest and lowest values with their counts (can also chain to head() here).

In [9]:
print("Unique Odometer Readings:", autos["odometer_km"].unique().shape[0],end='\n\n')
print(autos["odometer_km"].describe(), end='\n\n')
autos["odometer_km"].value_counts()

Unique Odometer Readings: 13

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64



odometer_km
150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: count, dtype: int64

There are `13` unique values in the `odometer_km` column. We also notice that most cars in this dataset have higher odometer_KM readings. Additionally, there don't appear to be any unrealistically high or low outliers in this column.

In [10]:
print("Unique Price Readings:", autos["price"].unique().shape[0],end='\n\n')
print(autos["price"].describe(), end='\n\n')
autos["price"].value_counts().sort_index().head(15)

Unique Price Readings: 2357

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64



price
0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
Name: count, dtype: int64

* There are `1421` entries of cars with a price tag of $0 which is obviously not valid. And since ebay is an auction site, it completely legitimate to have items starting at `$1`.


In [11]:
autos['price'].value_counts().sort_index(ascending = False).head(15)

price
99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
Name: count, dtype: int64

* Some of the cars are really heavily priced. Some of them are even priced over `$1 million`.

* We notice a sudden spike from `350,000` to `999,990` dollars, which differs significantly from the gradual increase in prices observed before. Given this anomaly, it's safe to exclude any data points exceeding `$350,000`.

In [12]:
# Filter the 'autos' dataset to include only rows where the price falls between $1 and $350,000
autos = autos[autos['price'].between(1, 350000)]

# Display descriptive statistics for the 'price' column after filtering
autos['price'].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

After filtering the dataset, the maximum value in the price column has become `350,000` instead of `99,999,999` dollars, which was highly unrealistic. Additionally, the mean price decreased from `9,840` to `5,889` dollars, reflecting the removal of extremely high-priced outliers.

## 6. Exploring the Date Columns

Let's now move on to the date columns and understand the date range the data covers. There are five columns that represent date values.

* `date_crawled` 
* `last_seen` 
* `date_created` 
* `registration_month` 
* `registration_year` 
  
In our dataset, the columns `date_crawled`, `date_created`, and `last_seen` are currently represented as strings. We'll convert them to a numeric format so we can understand them quantitatively. However, the `registration_month` and `registration_year` columns are already in numeric format.

Let's first understand how the values in the first three string columns are formatted. These columns all represent full timestamp values:


In [13]:
# Display the first 5 rows of the 'date_crawled', 'date_created', and 'last_seen' columns
autos[['date_crawled', 'date_created', 'last_seen']][0:5]

date_crawled         date_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

* It seems like dates in these columns are stored in following format, **YYYY-MM-DD HH:MM:SS**.
  

In [14]:
#percentage distribution of date_crawled
print("Unique Entries:", autos["date_crawled"].str[:10].unique().shape[0])
(autos["date_crawled"]
 .str[:10]
 .value_counts(normalize=True, dropna=False)
 .sort_index()
)*100

Unique Entries: 34


date_crawled
2016-03-05    2.532688
2016-03-06    1.404304
2016-03-07    3.601359
2016-03-08    3.329558
2016-03-09    3.308967
2016-03-10    3.218367
2016-03-11    3.257490
2016-03-12    3.691959
2016-03-13    1.566972
2016-03-14    3.654896
2016-03-15    3.428395
2016-03-16    2.960980
2016-03-17    3.162772
2016-03-18    1.291053
2016-03-19    3.477813
2016-03-20    3.788737
2016-03-21    3.737259
2016-03-22    3.298672
2016-03-23    3.222485
2016-03-24    2.934212
2016-03-25    3.160712
2016-03-26    3.220426
2016-03-27    3.109235
2016-03-28    3.486050
2016-03-29    3.409863
2016-03-30    3.368681
2016-03-31    3.183363
2016-04-01    3.368681
2016-04-02    3.547823
2016-04-03    3.860805
2016-04-04    3.648718
2016-04-05    1.309585
2016-04-06    0.317101
2016-04-07    0.140019
Name: proportion, dtype: float64

* It appears that the website was crawled daily over approximately a one-month period in `2016`, from `March` to `April`. The distribution of crawling dates is roughly uniform, with some fluctuations.

In [15]:
#percentage distribution of last_seen: The ad was removed after this date
#presumably because the car was sold
print("Unique Entries:", autos["last_seen"].str[:10].unique().shape[0])
(autos["last_seen"]
 .str[:10]
 .value_counts(normalize=True, dropna=False)
 .sort_index()
)*100

Unique Entries: 34


last_seen
2016-03-05     0.107073
2016-03-06     0.432410
2016-03-07     0.539483
2016-03-08     0.741275
2016-03-09     0.959539
2016-03-10     1.066612
2016-03-11     1.237517
2016-03-12     2.378256
2016-03-13     0.889529
2016-03-14     1.260167
2016-03-15     1.587563
2016-03-16     1.645218
2016-03-17     2.808607
2016-03-18     0.735097
2016-03-19     1.583445
2016-03-20     2.065273
2016-03-21     2.063214
2016-03-22     2.137342
2016-03-23     1.853186
2016-03-24     1.976732
2016-03-25     1.921137
2016-03-26     1.680222
2016-03-27     1.564913
2016-03-28     2.085864
2016-03-29     2.234119
2016-03-30     2.477093
2016-03-31     2.378256
2016-04-01     2.279419
2016-04-02     2.491506
2016-04-03     2.520334
2016-04-04     2.448265
2016-04-05    12.476063
2016-04-06    22.180583
2016-04-07    13.194688
Name: proportion, dtype: float64

The last three days contain a disproportionate amount of `last_seen` values. Given that these are `10` times the values from the previous days, it's unlikely that there was a massive spike in sales, and more likely that these values are to do with the crawling period ending and don't indicate car sales.

* Note: It is not possible for the crawler to see all the ads everyday!

In [16]:
print("Unique Entries:", autos["date_created"].str[:10].unique().shape[0])
(autos["date_created"]
 .str[:10]
 .value_counts(normalize=True, dropna=False)
 .sort_index()
)*100

Unique Entries: 76


date_created
2015-06-11    0.002059
2015-08-10    0.002059
2015-09-09    0.002059
2015-11-10    0.002059
2015-12-05    0.002059
                ...   
2016-04-03    3.885514
2016-04-04    3.685782
2016-04-05    1.181921
2016-04-06    0.325337
2016-04-07    0.125605
Name: proportion, Length: 76, dtype: float64

In [17]:
#distribution of registration_year
autos["registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

* `registration_year` ,minimum value is 1000 (way before cars were invented) and maximum value is 9999 (many years into the future) which is not possible.

## 7. Dealing with Incorrect Registration Year Data

As we mentioned earlier, the registration_year column contains incorrect entries, such as the minimum value of 1000, predating the invention of cars, and the maximum value of 9999, representing years too far in the future.

Since a car's registration year cannot exceed the listing's last_seen date, any vehicle with a registration year beyond 2016 is definitely inaccurate. Determining the earliest reasonable year is more challenging. Realistically, it might fall within the first few decades of the 1900s.



In [18]:
autos = autos[autos['registration_year'].between(1900, 2016)]
autos['registration_year'].value_counts(normalize=True).head(10)

registration_year
2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
2006    0.057197
2001    0.056468
2002    0.053255
1998    0.050620
2007    0.048778
Name: proportion, dtype: float64

* We can see that the majority of cars listed in the dataset were registered within the last `20` years.

##  8. Exploring Price by Brand

Exploring variations across different car brands is a common practice when analyzing car data. We can use aggregation to gain insights from the `brand` column.

In [19]:
brands = autos['brand'].value_counts(normalize=True)
brands

brand
volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
mitsubishi        0.008226
honda             0.007840
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002271
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001500
trabant           0.00

* `Volkswagen` has by far the most listings followed by `BMW` and `Opel`.

* It is evident that `German` manufacturers dominate the listings since top `five` spots are held by them.

* A lot of brands don't have a significant number of listings.

Since many brands don't have a significant percentage of listings, we'll focus our analysis on brands that represent more than 5% of the total car sales data.

In [20]:
# Extract brands with more than 5% of the total listings
most_common_brands = brands[brands > 0.05].index
most_common_brands

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object', name='brand')

In [21]:
# Store the mean prices of brands
brand_mean_prices = {}

# Iterate over the most common brands
for brand in most_common_brands:
    
    # Select data for the current brand, calculate its mean price, and store it to the dictionary
    current_brand = autos[autos['brand'] == brand]
    mean_price = current_brand['price'].mean()
    brand_mean_prices[brand] = round(mean_price)

# Display the mean prices of brands
brand_mean_prices

{'volkswagen': 5402,
 'bmw': 8333,
 'opel': 2975,
 'mercedes_benz': 8628,
 'audi': 9337,
 'ford': 3749}

The analysis highlights a price disparity among the top 6 brands in the car sales data:

* `Audi`, `BMW`, and `Mercedes Benz` vehicles are typically priced higher compared to their counterparts.
* `Ford` and `Opel` emerge as the most budget-friendly, while `Volkswagen` falls somewhere in between.

This pricing pattern may have contributed to the popularity of `Volkswagen` cars.


## 9. Storing Aggregate Data in a DataFrame

For the top 6 brands, we can use a similar method to obtain the average mileage for those cars, and then see if there's any link with the mean price.

In [22]:
brand_mean_mileages = {}

# Iterate over the most common brands
for brand in most_common_brands:
    
    # Select data for the current brand, calculate its mean mileage, and store it to the dictionary
    current_brand = autos[autos['brand'] == brand]
    mean_mileage = current_brand['odometer_km'].mean()
    brand_mean_mileages[brand] = round(mean_mileage)
    
# Display the mean mileages of brands
brand_mean_mileages

{'volkswagen': 128707,
 'bmw': 132573,
 'opel': 129310,
 'mercedes_benz': 130788,
 'audi': 129157,
 'ford': 124266}

In [23]:
brand_info = pd.DataFrame(data = {"mean_price":brand_mean_prices, "mean_mileage":brand_mean_mileages})
brand_info

mean_price  mean_mileage
volkswagen           5402        128707
bmw                  8333        132573
opel                 2975        129310
mercedes_benz        8628        130788
audi                 9337        129157
ford                 3749        124266

We can see that the more expensive brands generally tend to have a relatively higher mileage with the only exception being `Opel`.